lib/codegen-meta/src/cdsl/types.rs

```diff
@@ -0,0 +1,467 @@
+    }
+}
+
+/// Create a LaneType from a given bool variant.
+impl From<base_types::Bool> for LaneType {
+    fn from(b: base_types::Bool) -> Self {
+        let bits = b as u64;
+        let tag = LaneTypeTag::BoolType(b);
+        Self { bits, tag }
+    }
+}
+
+/// Create a LaneType from a given float variant.
+impl From<base_types::Float> for LaneType {
+    fn from(f: base_types::Float) -> Self {
+        let bits = f as u64;
+        let tag = LaneTypeTag::FloatType(f);
+        Self { bits, tag }
+    }
+}
+
+/// Create a LaneType from a given int variant.
+impl From<base_types::Int> for LaneType {
+    fn from(i: base_types::Int) -> Self {
+        let bits = i as u64;
+        let tag = LaneTypeTag::IntType(i);
+        Self { bits, tag }
+    }
+}
+
+/// Tags used to specify the kinds of elements in a lane type.
+#[derive(Debug, Clone, Copy)]
+pub enum LaneTypeTag {
+    BoolType(base_types::Bool),
+    FloatType(base_types::Float),
+    IntType(base_types::Int),
+}
+
+/// An iterator for different lane types.
+pub struct LaneTypeIterator {
+    bool_iter: base_types::BoolIterator,
+    int_iter: base_types::IntIterator,
+    float_iter: base_types::FloatIterator,
+}
+
+impl LaneTypeIterator {
+    /// Create a new lane type iterator.
+    fn new() -> Self {
+        Self {
+            bool_iter: base_types::BoolIterator::new(),
+            int_iter: base_types::IntIterator::new(),
+            float_iter: base_types::FloatIterator::new(),
+        }
+    }
+}
+
+impl Iterator for LaneTypeIterator {
+    type Item = LaneType;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let b @ Some(_) = self.bool_iter.next() {
+            b.map(LaneType::from)
+        } else if let i @ Some(_) = self.int_iter.next() {
+            i.map(LaneType::from)
+        } else if let f @ Some(_) = self.float_iter.next() {
+            f.map(LaneType::from)
+        } else {
+            None
+        }
+    }
+}
+
+/// A concrete SIMD vector type.
+///
+/// A vector type has a lane type which is an instance of `LaneType`,
+/// and a positive number of lanes.
+pub struct VectorType {
+    base: LaneType,
+    lanes: u64,
+}
+
+impl VectorType {
+    /// Initialize a new integer type with `n` bits.
+    pub fn new(base: LaneType, lanes: u64) -> VectorType {
+        VectorType { base, lanes }
+    }
+
+    /// Return a string containing the documentation comment for this vector type.
+    pub fn doc(&self) -> String {
+        format!(
+            "A SIMD vector with {} lanes containing a `{}` each.",
+            self.lanes,
+            self.base.name()
+        )
+    }
+
+    /// Get the name of this vector type.
+    pub fn name(&self) -> String {
+        format!("{}X{}", self.base.name(), self.lanes,)
+    }
+
+    /// Find the unique number associated with this vector type.
+    pub fn number(&self) -> u8 {
+        let b = f64::from(self.base.number());
+        let l = (self.lanes as f64).log2();
+        let num = 16_f64 * l + b;
```

Comment:
Can we do this computation without using floating point? I guess it's a little more work because we need a helper:

```rust
fn floor_log2(x: u64) -> u32 {
  63 - x.leading_zeros()
}
```
but it would mean that we won't need to add an exception for the `float_arithmetic` clippy lint that we enable in some configurations.

Also, we should give the "16" constant a name.